### ISSUE TITLE
Zotero 7

### ISSUE THREAD

**Issue by wshanks** [wshanks/Zutilo#248]
From the [Zotero developer documentation](https://www.zotero.org/support/dev/zotero_7_for_developers):

> All Zotero plugins will need to be updated for Zotero 7.
> 
> Zotero 7 plugins continue to be based on the legacy XUL/XPCOM Mozilla extension framework that provides full access to platform internals, but the Mozilla platform itself no longer supports this framework. All Firefox extensions are now based on the much more limited WebExtensions API shared with Chrome and other browsers, which provides sandboxed APIs for common integration points.
> 
> We have no plans to make similar restrictions in Zotero. However, due to the Mozilla platform changes, some integration techniques are no longer available, and all plugins will need to change the way they register themselves in Zotero.

I haven't tried to scope out how much work this requires. Luckily Zutilo is already a bootstrap extension so some of the transition steps have already been done. @qqobb do you have any interest in working on this? Since there has not been much activity in the repo recently, I think it would be simplest just to do a clean transition and not try to support 6 and 7 in the same branch.

**Comment by qqobb** [comment]
I haven't looked into the Zotero 7 update either. I can try to help with the transition, but I'm not sure how much I'll be able to do. Let's see if there's a good way to collaborate.

There are also a few minor (old) commits which I could still add to the Zotero 6 branch of Zutilo. I'll try to create related PRs this weekend.

**Comment by SR--** [comment]
Just to say that it would be great to have it compatible with v7

**Comment by redleafnew** [comment]
Will it be compatible with zotero v7? Zotero 7 beta was released.

**Comment by wshanks** [comment]
I made an attempt at following the [plugin transition guide for Zotero 7](https://www.zotero.org/support/dev/zotero_7_for_developers) in the [zotero7 branch](https://github.com/wshanks/Zutilo/tree/zotero7). The current status is that it is possible to install the xpi file. However, some of the Zutilo code uses deprecated API's, so besides seeing Zutilo show up in the preferences pane and the Addons Manager all it does is produce errors right now. Definitely do not install it in a profile you care about because it currently makes the Tools menu unusable which then makes it difficult to open the Addons Manager and remove it.

In the best case scenario, there are just some deprecated function calls that be easily migrated to new functions. It is possible though that Zutilo could rely on something that does not have an easy replacement and more work is needed to update it. The next step is to run Zutilo, check the error console, address the error, and repeat until it gets to a usable state or a bigger issue is discovered. I can't make any guarantees about a timeline for getting Zutilo to work with Zotero 7. Others are welcome to try doing more work on the update if they want to.

**Comment by wshanks** [comment]
I played with it some more. I fixed the mangling of the Tools menu, so now it is safer to test because you can access the menu to remove it.

I also updated some other deprecated function usage. Now, the default item menu items (copy tags, paste tags, copy item, relate items) are all working. The preferences pane is not working. The next thing to do is get the preference pane working so that the other menu items / shortcuts can be tested.

**Comment by redleafnew** [comment]
Thanks for your hard working.

**Comment by wshanks** [comment]
> There was some interesting discussion regarding issues with Zotero7 transition on BetterBibtex repository, migh be useful; Is there any beta release to try and test?

No, I made one push of working on it back in June, but I have not come back to it so far. There is a `zotero7` branch in the repo that has my latest attempts, but it is not working. I had asked some questions on the zotero-dev mailing list and got some replies, but I hadn't used those replies to get further so far.

**Comment by xuxinliang** [comment]
Thanks for your hard working.

**Comment by topps** [comment]
I certainly appreciate all your hard work on this and also how much more difficult it will be to fix, given that the framework is now very much more restricted. 
To take the pressure off a little (and I hope this does not come across as unappreciative of your efforts), I have found a workaround for the most important function that Zutilo provided for me: copy item link. 
This functionality is now also provided by https://github.com/windingwind/zotero-actions-tags and is compatible with Zotero 7. It is a bit finicky to work with the provided Javascript action scripts, and some will be uncomfortable with allowing 3rd party Javascript to interact directly with their machine. But it does work. 
Thanks again for all your hard work.

**Comment by danieltomasz** [comment]
hi @topps if you need just (markdown) link, you can try use a markdown zotero translator as a workaround https://github.com/silentdot/zotero-markdown-translator/tree/main

**Comment by justinbrbooth** [comment]
Has there been any progress on this? I'm looking at switching to zotero for reading papers and being able to use proper keyboard navigation would be a huge plus

**Comment by bjohas** [comment]
Zotero 7 has been released. @wshanks: Can I help with upgrading somehow?

**Comment by SR--** [comment]
Yeah, I tried installing on Z7, but it seems to be disappearing on Z7 restart.

**Comment by bjohas** [comment]
Yes, plugins don't work for z7 by default but need to be explicitly upgraded....

**Comment by SR--** [comment]
Naturally, I have used the 'addon' content from the 7 branch as advised. The plugin loads OK but then does not reload on Zotero restart.

**Comment by joelparham** [comment]
Thanks for your continued development work. As a heavy user of Zotero, I am eager to have a Z7 compatible Zutilo as soon as possible.

**Comment by wshanks** [comment]
@bjohas Yes, any help would be appreciated. On the `zotero7` branch, I have done enough to get the plugin to load and have some functionality. It is able to add items to the item context menu and some of them work at least. I haven't tested all items, but I noticed at least paste tags did not work because it could not read from the clipboard but copy tags did work. The preferences window/pane is currently not accessible. It would be nice to fix that to make testing and using the plugin easier (but one can edit the preferences directly in the Firefox config editor). In a quick test, keyboard shortcuts did not seem to work but I haven't looked at them much yet (it would be nice to get the preferences UI working first).

**Comment by llity** [comment]
https://github.com/windingwind/zotero-plugin-template Using this framework makes development easier?

**Comment by wshanks** [comment]
I don't think that template helps because Zutilo is already written. https://www.zotero.org/support/dev/zotero_7_for_developers has a good guide for updating a plugin from Zotero 6 to Zotero 7. It just takes time to work through the changes.

I just pushed a change that gets the item menu preferences to appear in the Zutilo pane of the preferences window. It doesn't look great but it is functional and should make it easy to test which functions work. The main remaining steps to take for the Zotero 7 update are:

1. Test each function to see if it still works and fix the broken ones (some might not make sense any more with the changes to Zotero).
2. Convert the dtd files to ftl so that the strings appear properly in the preferences pane.
3. Turn on the keyboard shortcut code (just commented out right now) and see if it works.

**Comment by bjohas** [comment]
@wshanks - I'll have a look. It won't be fast, but with some colleagues we'll do some testing and see what we can fix.

**Comment by Dcrizz** [comment]
I managed to get this to work for copying and pasting tags for any item

### Copy Tags


`
// Copy tags.
// Codes from Zutilo.
// 获取条目tags，并存入剪贴板
var tagsArray = [];
var items = Zotero.getActiveZoteroPane().getSelectedItems();
for (let item of items) {
    // Check if the item is not a collection
    if (!item.isAttachment() && !item.isAnnotation()) {
        var tempTags = item.getTags();
        for (var j = 0; j < tempTags.length; j++) {
            let tag = tempTags[j].tag;
            if (!tagsArray.includes(tag)) {
                tagsArray.push(tag);
            }
        }
    }
}
var clipboardText = tagsArray.join('\r\n');
Zotero.Utilities.Internal.copyTextToClipboard(clipboardText);
return "Tags copied.";
`


### Paste Tags:

`
// Paste tags
// Codes from Zutilo.
// 从剪贴板获取tags
var clipboardText = Zotero.Utilities.Internal.getClipboard("text/plain").trim();
if (!clipboardText) {
    return false;
}
// 拆分
var tagArray = clipboardText.split(/\r\n?|\n/);
tagArray = tagArray.map(function (val) { return val.trim(); });
tagArray = tagArray.filter(Boolean);
// 添加tags
var items = Zotero.getActiveZoteroPane().getSelectedItems();
for (let item of items) {
    // Check if the item is not a collection
    if (!item.isAttachment() && !item.isAnnotation()) {
        for (let tag of tagArray) {
            item.addTag(tag);
        }
        item.saveTx();
    }
}
return "Tags pasted.";
`

Hope that helps

**Comment by wshanks** [comment]
Thanks, @dcrizz! I used your paste code in d89ad53. I didn't use the copy code since copy was already working. I am not sure if it is safe to rely on `Zotero.Utilities.Internal` though -- maybe we need to copy the code into Zutilo.

@bjohas Thanks -- if you want, you can post any progress as PRs like another user did in #270. I will keep chipping away at things as well and pushing to the `zotero7` branch.

**Comment by wshanks** [comment]
I made some more changes to the `zotero7` branch. Now all of the context menu items work like they did with Zotero 6. There is one issue I am trying to fix still -- some of the strings in the preferences pane always show in English instead of the translation when it is available.

I might make a new release with this functionality for Zotero 7. I haven't looked much at the keyboard shortcuts yet. They might require more work to update. The shortcut code is based on the old KeyConfig Firefox addon. I have used that in other contexts and had to switch to something else in higher versions of Firefox like Zotero is using now.

**Comment by wshanks** [comment]
@TommyJazz93 So Zutilo is interfering with pdf shortcuts in Zotero 7? Strange -- something to look into more while working on the Zotero 7 updates.

**Comment by Dcrizz** [comment]
> Thanks, @Dcrizz! I used your paste code in [d89ad53](https://github.com/wshanks/Zutilo/commit/d89ad53ee1facf14ca4ff6966cf3551f05af6100). I didn't use the copy code since copy was already working. I am not sure if it is safe to rely on `Zotero.Utilities.Internal` though -- maybe we need to copy the code into Zutilo.
> 
> @bjohas Thanks -- if you want, you can post any progress as PRs like another user did in #270. I will keep chipping away at things as well and pushing to the `zotero7` branch.

Sorry for the late reply!! No worries, to be honest, I don't know what I am doing really, I just messed around with it the original action tag a bit and got GPT to correct any errors I made!! I am not actually very competent at this stuff!!

**Comment by regis586** [comment]
Hi,
 zutilo is indispensable to me :) As it is not - for the moment - compatible with Zotero 7, I've kept version 6.0.36 "under wraps" to continue using it. But this morning the same incompatibility message appeared with 6.0.36, and indeed it doesn't work any more? No doubt linked to the future version for Zotero 7.x? 
Good luck to the developers for their excelent work...

**Comment by regis586** [comment]
I deleted the extension and reinstalled it with Zotero 6.0.36..and  zutilo works (fine..) again...

**Comment by glennsandstrom** [comment]
I use Zutilo for tag management all the time and It would be great if a version working on Zotero 7 becomes avalaible. Thanks for all the good work.

Best,

Glenn

**Comment by mmyrte** [comment]
Hi @wshanks - thanks for Zutilo! If I were you, I'd release a version as it stands, since the following menu items seem to work:

- Copy tags
- Remove tags
- Paste tags
- Relate items
- Copy attachment paths
    - As opposed to the "Show attachment paths", this gives me the absolute paths of files, even under WebDAV management.
- Copy creators
- QuickCopy items
    - Might want to update the docs from "Default output format" to "Item format and Note Format, respectively" in case someone's looking for the exact string.
- Copy select item links
    - I must have fucked something else up in my Zotero though, because I cannot actually follow the links. They look correct though.
- Copy select collection link
- Copy Zotero URIs + Open Zotero URIs
- Copy Zotero URI (collection)
- Create book section
- Create book item
- Copy item fields
- Paste into empty item fields
- Paste non-empty item fields
- Paste all item fields
- Paste item type

Unsure about the following / might not be working:

- Show attachment paths (Named "Show attachments" in settings)
    - Simply shows `Zutilo: showing path of attachment 1 of 2 / storage:index.html`. Is this the intended function? I'm using webdav storage.
- Modify attachment paths (Named "Modify attachments" in settings)
    - Likely also not working because of webdav; I'm getting `Zutilo: item selection error`
- Modify attached URLs
    - Only seems to work with checkbox active for "Replace all instances"
- Copy child items to relocate / Relocate copied child items
    - If I understand this correctly, it's supposed to copy the attachments on an item and move them to another item. The name "relocate" is ambiguous, is this a copy or move operation? In any case, it doesn't do anything I could detect.

Not tested

- QuickCopy items to clipboard (alt 1. and 2.)
    - I don't have the time to set this up, sorry.

I would suggest that you rename the plugin from "Zutilo Utility for Zotero" to simply "Zutilo" in the `manifest.json`, because I'm getting an overflow in preferences:

![image](https://github.com/user-attachments/assets/ad589b09-4bbd-4402-adb7-2a65cdbe5145)

**Comment by AshtoBashto** [comment]
While I wait for the 7.0 compatible version of zutilo... is there a way to get a zotero://select link from a zotero document in any other way? I'm pretty desperate

**Comment by schoukah** [comment]
@AshtoBashto I'm not sure where I found mention of this plugin but this does the trick for me while waiting for a compatible Zutilo release: https://github.com/windingwind/zotero-actions-tags#readme

**Comment by jan-glx** [comment]
I just tested QuickCopy items to clipboard (alt 1 and 2) with the current 7er branch. It works fine. If you dont want to wait for release:
[dowload a zip *of the contents* of the addon directory of the zotero7 branch](https://download-directory.github.io/?url=https%3A%2F%2Fgithub.com%2Fwshanks%2FZutilo%2Ftree%2Fzotero7%2Faddon), rename as .xpi & install


Link goes to a 3rd party tool that provides downloads of github repo directories [as this is not provided by github](https://stackoverflow.com/q/7106012/1870254)
Instructions based on : https://github.com/wshanks/Zutilo/blob/zotero7/docs/DEVELOPERS.md

**Comment by bmhj2** [comment]
> I just tested QuickCopy items to clipboard (alt 1 and 2) with the current 7er branch. It works fine. If you dont want to wait for release: [dowload a zip _of the contents_ of the addon directory of the zotero7 branch](https://download-directory.github.io/?url=https%3A%2F%2Fgithub.com%2Fwshanks%2FZutilo%2Ftree%2Fzotero7%2Faddon), rename as .xpi & install
> 
> Link goes to a 3rd party tool that provides downloads of github repo directories [as this is not provided by github](https://stackoverflow.com/q/7106012/1870254) Instructions based on : https://github.com/wshanks/Zutilo/blob/zotero7/docs/DEVELOPERS.md

Really pleased to find this - thank you. The loss of Zutilo has been the one thing making me sad since the Z7 update, so thanks @wshanks, @qqobb, @jan-glx, and others in this thread who've been working on it. Greatly appreciated.

**Comment by huachuman** [comment]
Interesting @jan-glx. Do you know a way to assign shortcut to, for example, quick copy item link? Currently the Shortcuts section seems to be empty and/or unresponsive...

**Comment by jan-glx** [comment]
I did not notice/read about that copy issue. -But its awesome! Its a feature, not a bug! Now you can copy the citekey when having the paper open in a build-in viewer tab - no need to switch back to the list or manually copy from the item pane 🚀
Copy by rightclick-> copy still works in case you need it and you can set another shortcut such that Zutilo does not steal ctrl+C

I had set the shortucts with the previous version, thus did not notice the glitch in the GUI
but instead you can use: Zotero->Edit->Settings->Advanced->Config editor
to set my shortcuts
![image](https://github.com/user-attachments/assets/9dcd4964-67ea-480f-929b-ece8a641190c)

where the ids can be derived from extensions.zotero.export.quickCopy.setting as explained elsewhere

**Comment by TommyJazz93** [comment]
@SR-- huachuman, yes, the issue is still present with that link. Sorry guys, maybe I'm too noob, but I still have the problem of no copy shortcut on pdf viewer when zutilo plugin is activated. I didn't get if there is a solution to that, and what is that solution. Answer would be really appreciated

**Comment by sypianski** [comment]
Thank you for your work, I can't wait to be able to use it.

**Comment by DJYeh** [comment]
Am I understanding correctly that this plugin is currently not compatible with the latest version of Zotero? If so I suggest updating the README file. I spent almost an hour trying to troubleshoot before I found this post.

**Comment by 01tot10** [comment]
> Am I understanding correctly that this plugin is currently not compatible with the latest version of Zotero? If so I suggest updating the README file. I spent almost an hour trying to troubleshoot before I found this post. 

@DJYeh, just a regular user posting here, but I don't think you've wasted your time, as many users including me have been able to use the development version succesfully with Zotero 7. As far as I know, the feature set is still not at parity with the earlier Zotero version, but I've found everything that I need even from the current development version. Maybe skim over this thread to get an idea of the limitations, as well as for the install instructions. You are right, though, that the current state of affairs could be reflected more clearly e.g. at the README.md.

**Comment by joelwwh** [comment]
Is there a way to get 'copy select link' to go straight to the correct sub-folder that stores the doc rather than just defaulting to the main library?

**Comment by ryanwwest** [comment]
I doubt it as there can be multiple sub-folders in which the corresponding item is stored. I don't know of any functionality to specify a particular sub-collection 'path'.

**Comment by yanisallouch** [comment]
> > Am I understanding correctly that this plugin is currently not compatible with the latest version of Zotero? If so I suggest updating the README file. I spent almost an hour trying to troubleshoot before I found this post.
> 
> @DJYeh, just a regular user posting here, but I don't think you've wasted your time, as many users including me have been able to use the development version succesfully with Zotero 7. As far as I know, the feature set is still not at parity with the earlier Zotero version, but I've found everything that I need even from the current development version. Maybe skim over this thread to get an idea of the limitations, as well as for the install instructions. You are right, though, that the current state of affairs could be reflected more clearly e.g. at the README.md.

I agree that the ` README`  should reflect that the release version is not working with Zotero 7.x more clearly.

**Comment by acsr** [comment]
I updated Zotero to Zotero 7.1-beta.3+52c342b69 due to fix some selection issues in highlighting of text in snapshots.

In the Addon Manager I now got the error:
"Zutilo Utility for Zotero 3.10.1a1 is incompatible with Zotero 7.1-beta.3+52c342b69."

In the manifest.json there is a version pinning limiting to 7.0.:
"strict_max_version": "7.0.*"

Changing this string manually to
"strict_max_version": "7.*"
did not act as a workaround.

There is no qualified error what exactly is failing beneath the now "dirty fixed" version check.

I am aware of the purpose of this check and just want to report it for those using the "Zotero7 branch addon folder zip to xpi" hack. Before the interims version worked well except the keyboard shortcuts.

Question: Is there anything else necessary fo fix for 7.1.* Where should I look to debug addon install logs?
Is this worth a new ticket?

**Comment by thynus** [comment]
Thanks for posting this!  I just did a copy-tags paste-tags and it worked!  I will keep an eye on the Mozilla error console as I am using it so I can share.

There were 4 of these:
[zoteroOverlay.js:339:57](chrome://zutilo/content/zoteroOverlay.js)

Uncaught TypeError: Zotero.Utilities.Internal.getClipboard(...) is null
    clipboard chrome://zutilo/content/zoteroOverlay.js:339
    pasteJSON chrome://zutilo/content/zoteroOverlay.js:1017
    pasteJSONIntoEmptyFields chrome://zutilo/content/zoteroOverlay.js:1021
    refreshZoteroPopup chrome://zutilo/content/zoteroOverlay.js:1059
    refreshZoteroItemPopup chrome://zutilo/content/zoteroOverlay.js:24

**Comment by wshanks** [comment]
I went ahead and merged the `zotero7` branch into `master` and tagged [release 4.0.0](https://github.com/wshanks/Zutilo/releases/tag/v4.0.0). I haven't been able to put more time into figuring out why the translations don't work completely or how to get the keyboard shortcuts working, but I hope having a release helps some people since it seems like several have been working off builds from the `zotero7` branch. Also, I had wanted to test if updates still work but hadn't gotten to that so far.

I also added a note to the readme to help indicate that the project is having trouble keeping up with updates to Zotero.

**Comment by acsr** [comment]
> I went ahead and merged the `zotero7` branch into `master` and tagged [release 4.0.0](https://github.com/wshanks/Zutilo/releases/tag/v4.0.0).

@wshanks Thanks for the initial official troubleshooting release.

I could load Zutilo  4.0.0 into Zotero 7.1-beta.11+fb8b65b32 on macOS Sequoia 15.3, Apple Silicon over an existing disabled Zutilo Quick&Dirty Build from the zotero7 branch. 
- Install was using the `Check for Updates` menu in the PlugIns Manager.
- There were no errors
- The prefs or menu items do not show up at any location
- Even after a restart no UI features or error messages

Until now I did not :
- activate Logging or Restart in Troubleshooting Mode.
- remove the old install and install from scratch (to avoid loosing prefs …)
- Test the 4.0.0 on a regular latest stable 7.0.x 

You may add Alpha state to reflect the instable state, as if it currently is not working at all anymore on early 7.1

**Comment by wshanks** [comment]
Thanks, @acsr, I debugged a couple issues with Zotero 7.1 and made a new 4.0.1 release of Zutilo to address them.

This also gave me a chance to test the automatic updates and I found that Zotero did update Zutilo 4.0.0->4.0.1, so they seem to be working.

**Comment by acsr** [comment]
@wshanks, great quick move!
- Update to 4.0.1 worked in 7.1-beta.11+fb8b65b32 on macOS Sequoia 15.3, Apple Silicon by calling `Check for updates`
- Prefs are back and preserved
- Menus are visible (not tested deeper in functionality)
- Shortcuts not tested, due to need to lookup my docs of the old preserved setup. (I forgot them due to missing the daily Zutilo KATA for month).

**Comment by regis586** [comment]
Hi !
Thanks a lot..
zutilo's born again.. :-)
It works fine (..anyway, for me..) with Zotero 7.015 (64 bits)..

**Comment by POBrien333** [comment]
@wshanks 
This is issue is also probably ready to be closed.

**Comment by bmhj2** [comment]
> [@wshanks](https://github.com/wshanks) This is issue is also probably ready to be closed.

Yes, though the same issue could be raised in relation to Zotero 10 (sorry!)

**Comment by POBrien333** [comment]
That was done already and I also provided a PR: https://github.com/wshanks/Zutilo/issues/309

**Comment by bmhj2** [comment]
> That was done already and I also provided a PR: [#309](https://github.com/wshanks/Zutilo/issues/309)

Legend.

**Comment by paregorios** [comment]
@bmhj2 and @POBrien333 please see https://github.com/wshanks/Zutilo/issues/309#issuecomment-5358278857 (happy to be corrected)
